crates/loader/src/local/mod.rs

```diff
@@ -361,6 +368,70 @@ impl ComponentDigest {
     }
 }
 
+/// The parsing of a `component.trigger` table depends on the application trigger type.
+/// But serde doesn't allow us to express that dependency through attributes.  In lieu
+/// of writing a custom Deserialize implementation for the whole manifest, we instead
+/// leave `component.trigger` initially unparsed, then once we have the rest of the
+/// manifest loaded, we go back and parse it into the correct enum case - we basically
+/// clone the partially-parsed manifest but changing the type of `component.trigger`
+/// (which is why this can't be a straight clone).  The next few functions accomplish
+/// this.
+///
+/// (The reason we can't continue using an untagged enum is that an external trigger
+/// might have a setting called `route` which would make serde parse it as a
+/// TriggerConfig::Http.  There are other ways around this, e.g. the trigger match
+/// checker could see External/Http and transform the typed config back to a HashMap -
+/// I went this way to avoid having to know about individual types but it has its
+/// own downsides for sure.)
+fn resolve_partials(
+    partially_parsed: RawAppManifestAnyVersionPartial,
+) -> Result<RawAppManifestAnyVersion> {
+    let RawAppManifestAnyVersionPartial::V1(manifest) = partially_parsed;
+
+    let app_trigger = &manifest.info.trigger;
+    let components = manifest
+        .components
+        .into_iter()
+        .map(|c| resolve_partial_component(app_trigger, c))
+        .collect::<Result<_>>()?;
+
+    Ok(RawAppManifestAnyVersion::V1(RawAppManifest {
+        info: manifest.info,
+        components,
+        variables: manifest.variables,
+    }))
+}
+
+fn resolve_partial_component(
+    app_trigger: &ApplicationTrigger,
+    partial: RawComponentManifestPartial,
+) -> Result<RawComponentManifest> {
+    let trigger = resolve_trigger(app_trigger, partial.trigger)?;
+
+    Ok(RawComponentManifest {
+        id: partial.id,
+        source: partial.source,
+        description: partial.description,
+        wasm: partial.wasm,
+        trigger,
+        build: partial.build,
+        config: partial.config,
+    })
+}
+
+fn resolve_trigger(
+    app_trigger: &ApplicationTrigger,
+    partial: toml::Value,
+) -> Result<TriggerConfig> {
+    use serde::Deserialize;
+    let tc = match app_trigger {
+        ApplicationTrigger::Http(_) => TriggerConfig::Http(HttpConfig::deserialize(partial)?),
```

Comment:
While this was not added by this PR, I wonder if at some point, we can rename `TriggerConfig` to `ComponentTriggerConfig`. `ApplicationTrigger` seems aptly named as it defines the trigger type for an application. However `TriggerConfig` does not have that level of specificity. Also, the enum options for each can be confusing: 
```
pub enum ApplicationTrigger {
    /// HTTP trigger type.
    Http(HttpTriggerConfiguration),
    /// Redis trigger type.
    Redis(RedisTriggerConfiguration),
    /// A trigger type that is not built in.
    External(ExternalTriggerConfiguration),
}

pub enum TriggerConfig {
    /// HTTP trigger configuration
    Http(HttpConfig),
    /// Redis trigger configuration
    Redis(RedisConfig),
    /// External trigger configuration
    External(HashMap<String, String>),
}
```